modules/sdk/gradle-plugins-node/CHANGELOG.markdown

```diff
@@ -137,6 +137,18 @@ when compiling a Java project.
 - [LPS-73472]: The `NpmInstallTask`'s `nodeModulesCacheRemoveBinDirs` property
 is no longer available.
 
+## 3.1.0 - 2017-08-29
+
+### Added
+- [LPS-73070]: Add the ability to run `ExecuteNpmTask` instances concurrently
+even when pointing to a shared NPM's cache directory, if supported.
+- [LPS-73070]: By default, use the current user NPM's cache directory
+concurrently if running on NPM 5.
```

Comment:
@Ithildir Two questions on this one:

1. Did you mean to say `... the current user's NPM cache directory`? At the moment, it's referring to a user named NPM.
2. What are we concurrently running the user's NPM cache dir with?

Reply:
Gradle can run tasks in parallel, and so it can happen that multiple `npmInstall` tasks are running at the same time, for example one for each module in a multi-project build (like a Liferay Workspace).

NPM must be configured to point to a cache directory where it saves all its files (different from the `node_modules` directory inside a project). Before NPM 5, this cache directory wasn't designed to be accessed concurrently: only one NPM invocation at a time. Now, with NPM 5, multiple NPM invocations can access the same cache directory concurrently.

The Node Gradle plugin lets you choose where this NPM cache directory is, and this is how it decides the default value (that can be overwritten):

- if NPM < 5, every project has its own cache directory, so we're sure it won't be accessed concurrently
- otherwise, leave the default value for NPM, which is to point to the NPM cache in the user home dir

The plugin checks if the cache dir is outside the project dir. If yes, then its access is synchronized (so, if multiple `npmInstall` of different projects all run at the same time, then they will be forced to run one at a time). But now, if the plugin is configured to run with NPM >= 5, then by default this synchronization will be disabled.

Hope it makes sense 😅

Reply:
Thanks, @Ithildir! Crystal clear 👍